common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/NfcSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,170 @@
+                            request.cancel();
+                            clearPin(pin);
+                            mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+                            return;
+                        }
+                        mSmartcardCertBasedAuthManager.requestDeviceSession(new AbstractSmartcardCertBasedAuthManager.ISessionCallback() {
+                            @Override
+                            public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
+                                tryUsingSmartcardWithPin(pin, certDetails, request, session);
+                                clearPin(pin);
+                            }
+
+                            @Override
+                            public void onException(@NonNull final Exception e) {
+                                indicateGeneralException(methodTag, e);
+                                request.cancel();
+                                clearPin(pin);
+                                mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+                            }
+                        });
+                    }
+
+                    @Override
+                    public void onClosedConnection() {
+                        //Nothing needed.
+                    }
+                });
+                mSmartcardCertBasedAuthManager.startDiscovery(mActivity);
+            }
+        };
+    }
+
+    /**
+     * Checks to see if PIN for smartcard is correct.
+     * If so, proceed to attempt authentication.
+     * Otherwise, handle the incorrect PIN based on how many PIN attempts are remaining.
+     * @param pin char array containing PIN attempt.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @param session An ISmartcardSession created to help with interactions pertaining to certificates.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    protected void tryUsingSmartcardWithPin(@NonNull final char[] pin,
+                                            @NonNull final ICertDetails certDetails,
+                                            @NonNull final ClientCertRequest request,
+                                            @NonNull final ISmartcardSession session)
+            throws Exception {
+        final String methodTag = TAG + ":tryUsingSmartcardWithPin";
+        if (session.verifyPin(pin)) {
+            //If pin is successfully verified, we will use the certificate to perform the rest of the logic for authentication.
+            useSmartcardCertForAuth(certDetails, pin, session, request);
+            return;
+        }
+        final int attemptsRemaining = session.getPinAttemptsRemaining();
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
```

Comment:
actually.... maybe we can just rename onGetSessionFinished() to onPausedSmartCardDiscovery().

then... make that invoke stopDiscovery() in the NFC flow, and replace with none in USB flow (since the USB flow does NOT need to pause the discovery).

if we do this, I assume we should be able to pull this method up to the base class?

Reply:
(and we can add a comment in the NFC counterpart that we need to pause discovery because when the user enters pin, they will need to move the card away from the phone. keeping the discovery on might lead to unneccessary callback trigger)

Reply:
I'll still do the renaming because I think that makes sense, but this method still can't be pulled up to the base class with the suggested change alone because of the extra showPinDialog call in the NFC flow vs USB. This extra call is necessary because the NFC flow would be currently showing a different dialog (either "tap smartcard" or "keep steady", whatever the phrasing is for those) and it must go back to the pin dialog BEFORE setPinDialogErrorMode is called on the dialog holder. USB stays on the PIN dialog until the pin is correct or the flow cancels out. 

I think what I'll do is create some sort of similar extra logic method to hold the extra showPinDialog for NFC, maybe called "setPinDialogForIncorrectAttempt", and include the setPinDialogErrorMode method at the end in both (for completion, since it seems ... too custom to just call that method in the middle).  Both of the changes combined show allow the tryUsingSmartcardWithPin method to be brought up.